### ISSUE TITLE
Chapter 7, Chapter 8

### ISSUE THREAD

**Issue by FritzPeleke** [ageron/handson-ml#514]
Hello Ageron, I have some few questions:
1) What is the idea behind using Kmeans as a preprocessing step in classification?
2) I tried training the MNIST dataset with the XGboost XGBClassifier and it ran hopelessly for long and it didn't come to an end. Is it because of the data size?
3) I also realised that in training the bagging classifier, one of my classifiers actually outperforms it slightly. Is it because the other two were poor and it had to compensate for them?
4) How can you make your ML models/programs run faster?

**Comment by ageron** [comment]
Hi @FritzPeleke ,

Thanks for your questions.

1. Using Kmeans as a preprocessing step results in a non-linear dimensionality reduction of your initial dataset which can help the downstream classifier in some cases.
2. Could you please share the code in a [gist](https://gist.github.com/)? Yes, it could be the data size, but I'm surprised because XGboost is pretty fast.
3. Yes, it's possible, especially if you only have 3 classifiers in the bag, and if the validation set is small. Adding more classifiers in the bag usually makes the ensemble better, but it's always possible that one of the individual classifiers will outperform the ensemble by pure chance on the validation set if the validation set is small.
4. One option is brute-force: use a computer with a faster CPU (and more RAM).  Another is to use an algorithm that parallelises well across multiple machines and/or processors (e.g., in Scikit-Learn, use a classifier that has an `n_jobs` hyperparameter).  You can also select faster algorithms (e.g., SGDClassifier rather than SVM for larger datasets). You can perform dimensionality reduction if the number of features is large. And if you build neural networks, you can train them on machines equipped with GPUs (or on Google Cloud TPUs).

Hope this helps.  :)

**Comment by FritzPeleke** [comment]
Hi Ageron, thanks for your replies. They helped.  I retrained the Xgboost and it worked. turns out I was importing it as 'from xgboost import XGBclassifier' so it didn't work but when I did 'from xgboost.sklearn import XCBClassifier' it worked. And I also reduced the MNIST dataset to just 1000.

Concerning neural networks, I am just about to start part two of your book. So I guess I will learn from there. However, I can tell you it is a wonderful book. Thank you for the knowledge in it.
